notify/impl.go

```diff
@@ -75,9 +75,9 @@ type integration interface {
 	name() string
 }
 
-// Build creates a fanout notifier for each receiver.
-func Build(confs []*config.Receiver, tmpl *template.Template) map[string]Fanout {
-	res := map[string]Fanout{}
+// BuildReceiverTree builds a tree of receivers and a unique id of notification providers.
+func BuildReceiverTree(confs []*config.Receiver, tmpl *template.Template) map[string]map[string]Notifier {
+	res := make(map[string]map[string]Notifier)
 
 	filter := func(n integration, c notifierConfig) Notifier {
```

Comment:
This function actually contains another 2 stages and it's hard to see when they run. The filtering of resolved notifications and the actual sending+instrumentation (which should probably stay together for sanity.

The filtering can go into its own stage. I think it's even wrong here. As we account for resolved alerts on deduplication but throw them away later. So we might send the same notification anyway.